### PR TITLE
[fix](version) fix recover bug for lower version

### PR DESCRIPTION
**Pull request by nextdreamblue**
Signed-off-by: nextdreamblue <zxw520blue1@163.com>

# Proposed changes

relation with: #13067

## Problem summary

在 https://github.com/apache/doris/pull/13067 中对catalog recyble bin的优化中，在RecycleDatabaseInfo中增加了保存tableid的变量，用于支持根据id区分不同的同名table。但是这个信息在之前版本的元数据中是没有的，导致旧版本的fe升级到新的版本（VERSION_114）后，原有回收站中的database不能正确恢复，属于 https://github.com/apache/doris/pull/13067 pr代码的新旧版本的兼容性bug。

建议将该代码合并到包含 https://github.com/apache/doris/pull/13067 相关的分支中

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [x] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by nextdreamblue** [comment]
@morningman please review it

**Review by morningman — COMMENTED** [review]
Could you please add the number limitation of the recycled objects? Just as we discussed before. And the default number can be configured.

**Comment by nextdreamblue** [comment]
> Could you please add the number limitation of the recycled objects? Just as we discussed before. And the default number can be configured.

this issue will add in anothor pr

File: fe/fe-core/src/main/java/org/apache/doris/catalog/CatalogRecycleBin.java [hunk 1]
```diff
@@ -824,6 +824,32 @@ public void readFields(DataInput in) throws IOException {
...
+                    RecycleTableInfo tableInfo = entry.getValue();
+                    if (tableInfo.getDbId() != dbEntry.getKey()
+                            || !tableNames.contains(tableInfo.getTable().getName())) {
+                        continue;
+                    }
+
+                    tableIds.add(entry.getKey());
+                }
+                dbInfo.setTableIds(tableIds);
+                idToDatabase.put(dbEntry.getKey(), dbInfo);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/CatalogRecycleBin.java:850]: Why put it back to `idToDatabase`? it already exist in `idToDatabase`.

Comment (nextdreamblue) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/CatalogRecycleBin.java]: > Why put it back to `idToDatabase`? it already exist in `idToDatabase`.

fixed it

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.36 seconds
 load time: 437 seconds
 storage size: 17150491450 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221123010759_clickbench_pr_51110.html

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**